src/main/java/com/cdancy/bitbucket/rest/features/InsightsApi.java

```diff
@@ -0,0 +1,167 @@
+import com.cdancy.bitbucket.rest.parsers.RequestStatusParser;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.rest.annotations.BinderParam;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+import javax.inject.Named;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+
+@Produces(MediaType.APPLICATION_JSON)
+@RequestFilters(BitbucketAuthenticationFilter.class)
+@Path("/rest/insights/{jclouds.api-version}")
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public interface InsightsApi {
+
+    @Named("insights:get-annotations")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp2"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/annotations")
+    @Fallback(BitbucketFallbacks.AnnotationsResponseOnError.class)
+    @GET
+    AnnotationsResponse listAnnotations(@PathParam("project") String project,
+                                        @PathParam("repo") String repo,
+                                        @PathParam("commitId") String commitId,
+                                        @Nullable @QueryParam("externalId") String externalId,
+                                        @Nullable @QueryParam("path") String path,
+                                        @Nullable @QueryParam("severity") String severity,
+                                        @Nullable @QueryParam("type") String type);
+
+    @Named("insights:get-reports")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp4"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/reports")
+    @Fallback(BitbucketFallbacks.InsightReportPageOnError.class)
+    @GET
+    InsightReportPage listReports(@PathParam("project") String project,
+                                  @PathParam("repo") String repo,
+                                  @PathParam("commitId") String commitId,
+                                  @Nullable @QueryParam("limit") int limit,
+                                  @Nullable @QueryParam("start") int start);
+
+    @Named("insights:get-report")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp7"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/reports/{key}")
+    @Fallback(BitbucketFallbacks.InsightReportOnError.class)
+    @GET
+    InsightReport getReport(@PathParam("project") String project,
```

Comment:
:warning: Rather than using a lot of String arguments, consider using a container object for those values. (#QR12:88#)-PMD_UseObjectForClearerAPI